packages/utilities/src/FocusRectsProvider.tsx

```diff
@@ -0,0 +1,65 @@
+import * as React from 'react';
+import { FocusRectsContext } from './useFocusRects';
+
+export type FocusRectsProviderParams = {
+  /**
+   * Ref to the root element that this is providing focus rects for.
+   */
+  providerRef: React.RefObject<HTMLElement>;
+
+  /**
+   * Indicates that this is the root of a layer, and should not inherit the providerRef from a parent context.
+   */
+  layerRoot?: boolean;
+};
+
+export const FocusRectsProvider: React.FC<FocusRectsProviderParams> = params => {
+  const { providerRef, layerRoot } = params;
+  const [registeredProviders] = React.useState<HTMLElement[]>([]);
+  const parentContext = React.useContext(FocusRectsContext);
+
+  // Inherit the parent context if it exists, unless this is a layer root.
+  // This allows the topmost provider element in the DOM tree to handle the focus events.
+  // Since layers are in a separate HTML tree from their parent, they shouldn't use the parent's providerRef.
+  const inheritParentContext = parentContext !== undefined && !layerRoot;
+
+  const context = React.useMemo(
+    () =>
+      inheritParentContext
+        ? undefined
+        : {
+            providerRef,
+            registeredProviders,
+            registerProvider: (e: HTMLElement) => {
```

Comment:
nit: Typically I see `e` used for "event" and `el` used for "HTML element"

```suggestion
            registerProvider: (el: HTMLElement) => {
```